Cython/Utility/TypeConversion.c

```diff
@@ -290,15 +290,26 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
             return NULL;
         } else {
             return result;
         }
+    } else {
+        char* result;
+        Py_buffer view;
```

Comment:
Seeing some compile error suggesting `Py_buffer` is unavailable in some builds. Guessing this is related to Python Limited API support. To fix that looks like we want something like this

https://github.com/cython/cython/blob/bf59121ba5c9f79b4bbfeab05407c15e07e4cb15/Cython/Utility/StringTools.c#L767-L799

Reply:
Looking at that code, it doesn't adapt that well to here. The idea of `__Pyx_PyObject_AsStringAndSize` is to get a pointer to underlying memory in the object. However the other code creates a temporary `bytes` object, whose memory it uses within the lifespan of that temporary. Given `__Pyx_PyObject_AsStringAndSize` has no real way to keep temporary objects (like `bytes`) alive, passing pointers to said temporary objects back to users is UB. So we can't use that

Maybe there are some other options I'm overlooking. Though a simple solution would just be to error in the Python Limited API case

Reply:
> keep temporary objects (like `bytes`) alive

This also applies to buffers, though. The buffer view is only guaranteed to be usable until we call `PyBuffer_Release()`. After that, the buffer is allowed to become unavailable and should not be used any more.

Buffers need to be handled in place, where and as long as they are used.

Reply:
Yep agreed, which we do in this code currently

Am more meaning I don't see a good fallback for the Limited API on Python pre-3.11. Though could just be missing alternatives

If you have suggestions, please let me know 🙂

Reply:
Or actually does your comment ( https://github.com/cython/cython/issues/6750#issuecomment-2744947269 ) mean we should consider moving this code from `__Pyx_PyObject_AsStringAndSize` to the `string.from_py` functions?

Because then we could use the same fallback mechanism

Reply:
> consider moving this code from `__Pyx_PyObject_AsStringAndSize` to the `string.from_py` functions

Yes. It's safe to do there but not in `__Pyx_PyObject_AsStringAndSize`.

Reply:
The best general solution is probably `https://docs.python.org/3/c-api/arg.html#c.PyArg_ParseTuple` with `z#`,

If you look at the implementation of that it allows anything that implements the buffer protocol but doesn't specifically have `bf_releasebuffer` (which implies that the buffer is good so long as the object exists). That's "borrowed" but no worse than what we have for bytes.

Reply:
That's a good suggestion!

Haven't used the `PyArg_*Parse*` functions much before

Tried rewriting this below

Reply:
Have changed the logic to use `Py_buffer` where possible

Now `PyArg_Parse` is used as fallback

Reply:
> > consider moving this code from `__Pyx_PyObject_AsStringAndSize` to the `string.from_py` functions
> 
> Yes.

I think this got lost. It's still better to use the buffer protocol directly in places where we need the data within a well-defined scope. That's still more general than letting `__Pyx_PyObject_AsStringAndSize` exclude types that require buffer releasing.